pandas/tests/frame/methods/test_compare.py

```diff
@@ -257,8 +259,13 @@ def test_compare_ea_and_np_dtype(val1, val2):
             ("b", "other"): np.nan,
         }
     )
-    result = df1.compare(df2, keep_shape=True)
-    tm.assert_frame_equal(result, expected)
+    if val1 is pd.NA and is_numpy_dev:
+        # can't compare with numpy array if it contains pd.NA
+        with pytest.raises(TypeError, match="boolean value of NA is ambiguous"):
```

Comment:
Hmm why wouldn't we allow this? The returned array should just be object dtype with pd.NA no?

Reply:
comparing with an `object` ndarray which contains `pd.NA` now just raises https://github.com/pandas-dev/pandas/issues/50124#issuecomment-1342958334

Brock had commented here about this here, saying it'd be his inclination to say that raising is expected https://github.com/pandas-dev/pandas/pull/50129#discussion_r1043580061

Reply:
In this test, the column is being designed to be of dtype `object` with `pd.NA` in it:
```python
arr = [4.0, val1]  # val1 is pd.NA from the fixture
df1 = pd.DataFrame({"a": arr, "b": [1.0, 2]})
```
So I think `DataFrame.compare` is doing the right thing - `df1['a']` is an ordinary `object` array, not a masked one

Arguably it's the test that's doing something unsupported and should be rewritten so that `df['a']` is a nullable type (which can handle `pd.NA` just fine)

Reply:
Our extension arrays are 1D, as soon as you have to operate on more than one column, we coerce to object dtype. Not saying this is the case here, but this is inevitable in some places.

Reply:
Adding to this: the DataFrame in the test has dtype object, eg no masked array

Reply:
True, though fortunately the error's not come up in other places (yet! 😨 ). For EAs, if an error comes up as a result of this, then perhaps it's up to pandas to find a workaround

But for `object` arrays (non-EA) which contain `pd.NA`, do you agree that this is unsupported?

Reply:
I was mostly responding to the design flaw but. I totally agree that users shouldn’t do this

Reply:
> Our extension arrays are 1D, as soon as you have to operate on more than one column, we coerce to object dtype. Not saying this is the case here, but this is inevitable in some places.

This is a problem when processing masked arrays in Dataframe methods/functions. Maybe there could be made an internal dataframe-wide na mask method, so the we can deal with NA's properly on a dataframe level, for example in the `compare` method.

Reply:
> But for object arrays (non-EA) which contain pd.NA, do you agree that this is unsupported?

Yes, it's inherently usupported, as `bool(NA)` fails (and should fail). Seeing pd.NA in numpy array may be code smells that we should avoid and work with the masks instead.